src/cache/cache_las.c

```diff
@@ -536,26 +642,232 @@ __wt_las_remove_block(WT_SESSION_IMPL *session,
 
 	WT_STAT_CONN_DECRV(session, cache_lookaside_entries, remove_cnt);
+	__wt_cache_decr_check_uint64(session,
+	    &S2C(session)->cache->las_entry_count, remove_cnt,
+	    "lookaside entry count");
+	return (ret);
+}
+
+/*
+ * __wt_las_save_dropped --
+ *	Save a dropped btree ID to be swept from the lookaside table.
+ */
+int
+__wt_las_save_dropped(WT_SESSION_IMPL *session)
+{
+	WT_BTREE *btree;
+	WT_CACHE *cache;
+	WT_DECL_RET;
+
+	btree = S2BT(session);
+	cache = S2C(session)->cache;
+
+	__wt_spin_lock(session, &cache->las_sweep_lock);
+	WT_ERR(__wt_realloc_def(session, &cache->las_dropped_alloc,
+	    cache->las_dropped_next + 1, &cache->las_dropped));
+	cache->las_dropped[cache->las_dropped_next++] = btree->id;
+err:	__wt_spin_unlock(session, &cache->las_sweep_lock);
+	return (ret);
+}
+
+/*
+ * __las_sweep_init --
+ *	Prepare to start a lookaside sweep.
+ */
+static int
+__las_sweep_init(WT_SESSION_IMPL *session)
+{
+	WT_CACHE *cache;
+	WT_DECL_RET;
+	u_int i;
+
+	cache = S2C(session)->cache;
+
+	__wt_spin_lock(session, &cache->las_sweep_lock);
+	/* If no files have been dropped, there's nothing to do. */
+	if (cache->las_dropped_next == 0)
+		WT_ERR(WT_NOTFOUND);
+
+	/* Scan the btree IDs to find min/max. */
+	cache->las_sweep_dropmin = UINT32_MAX;
+	cache->las_sweep_dropmax = 0;
+	for (i = 0; i < cache->las_dropped_next; i++) {
+		cache->las_sweep_dropmin = WT_MIN(
+		    cache->las_sweep_dropmin,
+		    cache->las_dropped[i]);
+		cache->las_sweep_dropmax = WT_MAX(
+		    cache->las_sweep_dropmax,
+		    cache->las_dropped[i]);
+	}
+
+	/* Initialize the bitmap. */
+	__wt_free(session, cache->las_sweep_dropmap);
+	WT_ERR(__bit_alloc(session,
+	    1 + cache->las_sweep_dropmax - cache->las_sweep_dropmin,
+	    &cache->las_sweep_dropmap));
+	for (i = 0; i < cache->las_dropped_next; i++)
+		__bit_set(cache->las_sweep_dropmap,
+		    cache->las_dropped[i] - cache->las_sweep_dropmin);
+
+	/* Clear the list of btree IDs. */
+	cache->las_dropped_next = 0;
+
+err:	__wt_spin_unlock(session, &cache->las_sweep_lock);
+	return (ret);
+}
+
+/*
+ * __wt_las_sweep --
+ *	Sweep the lookaside table.
+ */
+int
+__wt_las_sweep(WT_SESSION_IMPL *session)
+{
+	WT_CACHE *cache;
+	WT_CURSOR *cursor;
+	WT_DECL_RET;
+	WT_ITEM *key, las_key;
+	uint64_t cnt, las_counter, las_pageid, remove_cnt;
+	uint32_t las_id, session_flags;
+	int notused;
+
+	cache = S2C(session)->cache;
+	cursor = NULL;
+	key = &cache->las_sweep_key;
+	remove_cnt = 0;
+	session_flags = 0;		/* [-Werror=maybe-uninitialized] */
+
+	__wt_las_cursor(session, &cursor, &session_flags);
+
+	/* We should have our own session. */
+	WT_ASSERT(session, cursor->session == &session->iface);
+
+	/*
+	 * If we're starting a new sweep, gather the list of trees to sweep.
+	 *
+	 * Otherwise, when continuing a sweep, position the cursor using the
+	 * key from the last call (we don't care if we're before or after the
+	 * key, either side is fine).
+	 */
+	if (key->size == 0) {
```

Comment:
The awkward nested if could be re-written as:
/*
 * If we are continuing a sweep, position the cursor...
 * Otherwise start a new sweep and gather the list of trees to sweep.
 */
if (key->size != 0) {
 <continue sweep code>
} else if ((ret == __las_sweep_init(session)) != 0)
   goto srch_notfound;